Improved filters

femans opened the pull request:
solves #162
This is a full refactor of the filter code. 
Filters are created as a filter definition, rather than inside a template + VM data + tightly coupled api/router code.
This will allow for:

- adding more/advanced filters, including nsfw filter
- adding custom filters
- type API query parameters in the URL of ipfs-search
- easier API maintenance

included new functionality is: 
- a filter for file extensions

BREAKING: The old URL queries do not work anymore. 
I found that it is rather unpractical and unnecessary to use a reflection of the API keys as queryparams for the filter selections. It seems to make more sense to encode the selections themselves in the query parameters.

TODO still: 

- [x] the file extension definitions based on type
- [x] refactor type filter; implemented the same way as the other filters (except different display)
- [x] change the query parameters to use a slug (like item.text) rather than the api keys

femans commented on .eslintrc.js:
```diff
@@ -40,4 +45,16 @@ module.exports = {
+  overrides: [
+    {
+      files: [
+        '**/__tests__/*.{j,t}s?(x)',
```
I added jest to the package, but didn't finally write any tests.

dokterbob replied:
Do you have any concrete proposals, regarding testing?

Is Jest really the best way to go about it? To be honest, I would personally suggest to look into something allowing us to write tests from a user's perspective, e.g. E2E or integration testing (like Selenium) in a browser, over anything like DOM-based testing. Also to take care of stuff like CSS hiding things based on viewport sizes etc.

Some resources:
* https://aswinkumar4018.medium.com/which-e2e-testing-framework-to-use-for-js-based-client-applications-fbcac9aab680
* https://www.netguru.com/blog/which-javascript-ui-testing-framework-to-use-in-2020

It seems that Good Old Selenium or Cypress are viable options.

It seems to me it would be wise contemplating this a little more before making a decision and, hence, leaving unused Jest-stuff out of this PR.

femans replied:
No, I was just happy to have jest installed so I can make unit tests. I did not want to remove it again so next for the next occasion I have to install it again. If you prefer, I will write a unit test to fix the bug you spotted so that it has more merit to stay? :D 

For e2e it seems puppetteer is good. 

https://blog.logrocket.com/react-end-to-end-testing-jest-puppeteer/

femans replied:
Now there is a test, so we stick with jest 👯‍♀️ Great if anyone cares to implement e2e testing

femans commented on src/components/results/list/mixins/FileListMixin.js:
```diff
@@ -41,7 +41,7 @@ export default {
     },
     pageCount() {
-      return Math.ceil(this.resultsTotal / batchSize);
+      return Math.ceil(store.getters[`results/${this.fileType}/resultsTotal`] / batchSize);
```
This fixes a bug that for some reason suddenly decided to become apparent.

femans commented:
@dokterbob @vandrongelen I added a filter for file extensions. But I did not yet add a lot of extensions. 
You can add your own favourite extensions to the filterDefinitions.js file.

BTW, the filter works well, except that many files don't have an extension, I noticed for pdf's. Not sure if it serves the purpose then so much.

dokterbob commented on src/components/SearchBar.vue:
```diff
@@ -92,7 +92,7 @@ export default {
+      get: () => store.getters['query/filters/typeFilter'].value,
       set(newType) {
         if (this.type !== newType) {
           enterSearchQuery({ type: newType });
```
Are we going to have code *exactly* like this for every filter we're adding?

In that case, I would suggest that *as soon as* we agree on the API, we abstract this somehow. Doesn't have to be part of the PR but an issue should be made *before* merging to keep track of it.

femans replied:
no, this is just the type filter. The type filter is the only non-regular filter. For the other filters you just have to add a json definition.

dokterbob commented on src/components/SearchFilters.vue:
```diff
@@ -22,63 +22,14 @@
-              </v-select>
+              <select-filter
+                :filter="filter"
+              />
```
Nice! Clean!

dokterbob commented on src/components/SearchFilters.vue:
```diff
@@ -88,61 +39,15 @@
 
 <script>
+import { computed } from '@vue/composition-api';
+import SelectFilter from '@/components/helpers/SelectFilter';
```
This seems an actual component, not a helper. Perhaps we should not have this in a helper folder, as thus far helpers are mere functions. What do you reckon'?

femans replied:
It is not the first component to end up in the helper folder. Feel free to improve the folder structure.

dokterbob replied:
Good enough for now.

dokterbob commented on src/components/helpers/filterDefinitions.js:
```diff
@@ -0,0 +1,215 @@
+  ],
+};
+
+export const extensions = {
```
I don't understand the relevance of extensions here if we have mimetypes defined, as well as an API/library for rendering those mimetypes as extensions for user-convenience. I would strongly recommend to filter based on mimetypes over extensions, also because most content has a mimetype while many files do not have even a filename let alone an extension defined.

I would also suggest to make the filtering API rich/expressive enough to:
1. Allow search results to define available type filters (so the list of types can be based on actual types in a search result).
2. Allow search queries to define available filters (e.g. directories won't have type filters, documents have language filters based on return languages etc.).

So then the list of filters would be a responsive thingy based on results and query from the store. Does that make sense?

dokterbob replied:
We might have to be careful to discriminate 'subtypes' (e.g. PDF, Word, ePub, MP3, MKV) from 'types' (e.g. Audio, Documents, Video) in the UX.

Perhaps best to discuss this with @vandrongelen.

femans replied:
yeah i agree this needs some more thought. 
It should be trivial to turn the extension filter into a mimetype filter.

femans replied:
> I don't understand the relevance of extensions here if we have mimetypes defined, as well as an API/library for rendering those mimetypes as extensions for user-convenience. I would strongly recommend to filter based on mimetypes over extensions, also because most content has a mimetype while many files do not have even a filename let alone an extension defined.

It might be better to make a mimetype-subfilter rather than extension filter indeed; the extension filter is not very good, because not all files have extensions. It is very easy to change the filter definitions to make it so.

> I would also suggest to make the filtering API rich/expressive enough to:
> 
> 1. Allow search results to define available type filters (so the list of types can be based on actual types in a search result).
> 2. Allow search queries to define available filters (e.g. directories won't have type filters, documents have language filters based on return languages etc.).
> 
> So then the list of filters would be a responsive thingy based on results and query from the store. Does that make sense?

1. Makes no sense. You can't prefilter the query based on its results. If you want to make post-query filter, this means rethinking the UX of it all. Currently, the filters update the query. 
2. This is already happening. If you select directories, there is no extension filter. The same mechanism can be used for other types of filters, e.g., if you wanted to make a language filter only for text documents.

dokterbob replied:
Alright, so let's remove extenion filters, or refactor them to use mimetypes.

dokterbob replied:
Also:

1. It seems that currently extension filters *are* disabled but I have not been able to discover where or how in the code this is done. I really need to know how to add more filters and I find the current code/API design *very* hard to understand. 

2. The current filter values do seem static, that is, they cannot arbitrarily change based on user-defined conditions

Couldn't it be as single as a `getFilters()`, which returns a bunch of queries (based on the query and the results from their respective stores)? The filters format might return a list of filters very similar to what `filterDefinitions.js` returns, although perhaps a slug-based hash makes composition more elegant and legible.

Note that making assumptions about types in the *architecture* if the filter infrastructure really seems an anti-feature, there is no guarantee that categories like `directory` and `text` will be maintained, or the concept of type categories in general for that matter. Plus it doesn't seem necessary to have this concept.

It seems to me it could be as simple as (pseudo-code based on imagination API):

```javascript
function getLanguageFilter() {
	let languages = getValuesFromResultsStore('language');
	...
}

function getSubtypeFilter() {
	let mimetypes = getValuesFromResultsStore('mimetypes');
	...

}

function getAnyFilters() {
	return {
		size: getSizeFilter(),
		...
	}

}

function getDocumentFilters() {
	return {
		subtype: getSubtypeFilter(),
		language: getLanguageFilter(),
		...
	}

}

function getFilters() {
	let type = getFilterFromQueryStore('type');

	switch (type) {
		case 'any':
			return getAnyFilters();
		case 'audio':
			return getDocumentFilters();
		...
	}

}
```

dokterbob commented on src/helpers/ApiHelper.js:
```diff
@@ -84,13 +39,11 @@ export function apiMetadataQuery(hash) {
 
   return api.searchGet(
-    query + typeFilter,
+    query,
```
I love this elegance.

dokterbob commented on src/store/modules/SearchQuery.js:
```diff
@@ -1,127 +1,42 @@
+    state.page = Number(routeParams.page) || defaultQuery.page;
+    Object.entries(state.filters).forEach(([slug, filter]) => {
+      filter.select(routeParams[slug]);
+    });
```
It would be great to have a little documented example of some filter and search queries here in a comment, that would greatly help with future maintenance as well as provide a basis for implementing tests.

dokterbob reviewed (changes requested):
Great work!

This is a big refactor, so let's reflect well on the API. Main thing is we should be able to use current query data (as AFAIK we already are in your PR) as well a results (e.g. what languages are present in results and filter defaults based on the selected type in the query) and in the future aggregation queries (remote API calls) to inform filter options. 

Another point is, it makes no sense to filter by extensions if we already have mimetype data (which is well-specified and we can already convert to extensions), although we should reflect carefully on UI relating to 'types' and 'subtypes'. I think we really need Kees' input on that.

In addition, we should consider our testing strategy, methodology and tooling carefully. Even without that decision, this PR is a heavy one so I would suggest to leave it out of this and discuss it separately.

Lastly, I came across some apparent search API/backend query errors:
[![image](https://user-images.githubusercontent.com/22742/157653546-333bf8b2-c3c5-4ddc-8204-67929e886e42.png)](https://bafybeih3f44aqwmuqxlblvxix5t3k2q7p2capiu55jx7yeebkmdr6lez6i.on.fleek.co/#/search?q=tree&page=1&size=10-100mb&type=text)

femans commented:
There is a bug still, in the main search page, there are no type filters applied for the separate queries. 
Somehow, this will have to be turned into an exception case, unfortunately.

dokterbob commented on src/store/index.js:
```diff
@@ -14,11 +15,11 @@ export default new Vuex.Store({
+        [Types.images]: SearchResults(Types.images),
+        [Types.audio]: SearchResults(Types.audio),
+        [Types.video]: SearchResults(Types.video),
+        [Types.directories]: SearchResults(Types.directories),
```
I really appreciate having these be proper constants!! :)

dokterbob commented on src/store/modules/SearchQuery.js:
```diff
@@ -1,127 +1,26 @@
+  modules: {
+    filters,
+  },
 };
```
Real great cleanup work.

Make sure to double check pagination!

dokterbob commented on src/store/modules/filterSubModule.js:
```diff
@@ -0,0 +1,150 @@
+    typeFilter: (state) => state.type,
+    mapFiltersToApi,
+  },
+};
```
Perhaps it's because I've never seen or used JS classes, but I find this code extremely hard to read. Also, there are quite a few (too many!) eslint exceptions.

At the very least, could you add a larger comment describing what does does, e.g. the kind of dataformat it will expect? Preferably also deal with eslint errors. Yes, I'm a style nazi - mostly about consistency though.

femans replied:
is this referring to the getters? Or the whole file?

dokterbob replied:
The whole module.

dokterbob reviewed (changes requested):
Real good work but I think the API could be a bit cleaner (as suggested before). See my simplified concept, over having static filters. If we make the API strong enough now we can really tweak filters at lightspeed and significantly improve the UX.

dokterbob commented on tests/mock-data.js:
```diff
@@ -0,0 +1,47 @@
+ * icecream is multiple select; you can have select 0 or more flavors
+ * @type {{mutations: {setValue}, state, getters: {toComponentProps, toSearchApi}}}
+ */
+const icecream = multipleSelectFilter({
```
It seems to me that this behaviour is essentially already implemented by the Vuetify [`<select>`](https://vuetifyjs.com/en/components/selects/) component. They also have a `text` and a `value`, [`multiple`](https://vuetifyjs.com/en/api/v-select/#props-multiple) selects and through [`value`](https://vuetifyjs.com/en/api/v-select/#props-value) the default can be set.

Is there any particular reason why you chose to implement this yourself? Something I'm not seeing?

femans replied:
This is a mock-module to test that data inserted into a module ends up properly in the components.

dokterbob reviewed (changes requested):
It seems to me there is still a lot of space to both clarify and simplify the API, but perhaps there's something I'm not seeing here.

It also seems we're re-implementing some functionality which is already available through Vuetify. Do you think we're perhaps better off using their code instead?

dokterbob commented on src/helpers/ApiHelper.js:
```diff
@@ -52,6 +52,17 @@ export function apiSearch(query, type, batch = 0, perBatch = batchSize) {
+        });
+      });
+      return results;
+    })
```
Nice nice nice nice!

femans commented on src/components/SearchBar.vue:
```diff
@@ -81,18 +81,21 @@ export default {
   computed: {
     searchPhrase: {
-      get: () => store.state.query.searchPhrase,
+      get() { return this.$store.state.query.searchPhrase; },
```
one less import, less chance of dependency cycles

femans commented on src/components/SearchBar.vue:
```diff
@@ -81,18 +81,21 @@ export default {
     type: {
-      get: () => store.state.query.type,
+      get() {
+        const single = (value) => [value].flat()[0];
```
yeah, i have been thinking about that. The point is that sometimes there is a string, and sometimes an array with length 1, and this helper resolves this.

femans commented on src/components/SearchBar.vue:
```diff
@@ -1,77 +1,77 @@
+        </template>
+      </v-text-field>
+    </div>
+  </v-container>
```
Now you can; with hide whitespaces. There are but 2 changes here, one is the test-id and the other is removal of an unused div tag.

femans commented on src/components/helpers/SelectFilter.vue:
```diff
@@ -0,0 +1,42 @@
+<script>
+import { enterSearchQuery } from '@/helpers/routerHelper';
+
+// TODO: make this a plugin of vue, to avoid the dependency
```
No, not sure. The reason for this idea is to remove the dependency, and thus remove chance of dependency cycles. They aren't a problem now though, and there are other ways of dealing with them if they come up.

femans replied:
Actually it is being used in a lot of places.

femans commented on src/store/modules/queryFilters/filterSubModule.js:
```diff
@@ -0,0 +1,65 @@
+    [TYPE]: typeFilter(typeFilterDefinition),
+    [LANGUAGE]: multipleSelectFilter(languageFilterDefinition),
+    [SIZE]: selectFilter(sizeFilterDefinition),
+    [LASTSEEN]: selectFilter(lastSeenFilterDefinition),
```
All requested functions, and unit tests, and filter-module-generator code is there now. But it is all built as Vuex-modules style getters and setters, and/or vuex-module-generator functions.

femans commented on src/store/modules/queryFilters/filterModuleGenerators.js:
```diff
@@ -0,0 +1,127 @@
+  label, slug, items, value, multiple,
+});
+
+export const selectFilter = (filterProperties) => filterModule({
```
wrote some comments above

femans commented:
@dokterbob I am happy with the code now. I still tested a bunch, fixed a few bugs, cleaned up some code, wrote some docstrings, repaired the unittests and resolved codeclimate issues since your last review/our call. I don't know if you want to look at it anymore, but for me it is ready to merge.
